File: app/src/main/java/com/hover/stax/bounties/BountyChannelsAdapter.java
package com.hover.stax.bounties;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.hover.stax.R;
import com.hover.stax.channels.Channel;
import com.hover.stax.views.StaxCardView;

import java.util.ArrayList;
import java.util.List;

class BountyChannelsAdapter extends RecyclerView.Adapter<BountyChannelsAdapter.CardedBountyListViewHolder> {

	private List<Channel> channelList;
	private List<Bounty> allBountiesList;
	private BountyListItem.SelectListener selectListener;

	public BountyChannelsAdapter(List<Channel> channels, List<Bounty> bounties, BountyListItem.SelectListener listener) {
		channelList = channels;
		allBountiesList = bounties;
		selectListener = listener;
	}

	@NonNull
	@Override
	public CardedBountyListViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
		View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.bounty_card_channel, parent, false);
		return new CardedBountyListViewHolder(view);
	}

	@Override
	public void onBindViewHolder(@NonNull CardedBountyListViewHolder holder, int position) {
		Channel c = channelList.get(position);
		holder.staxCardView.setTitle(c.getUssdName());
		List<Bounty> channelBounties = filterBounties(c.id);
		for (Bounty b: channelBounties) {
			BountyListItem bountyLi = new BountyListItem(holder.staxCardView.getContext(), null);
			bountyLi.setBounty(b, selectListener);
			holder.bountyListView.addView(bountyLi);
		}
	}

	private List<Bounty> filterBounties(int channelId) {
		List<Bounty> matches = new ArrayList<>();
		for (Bounty b: allBountiesList) {
			if (b.action.channel_id == channelId)
				matches.add(b);
		}
		return matches;
	}

	@Override
	public int getItemCount() {
		return channelList == null ? 0 : channelList.size();
	}

	static class CardedBountyListViewHolder extends RecyclerView.ViewHolder {
		private StaxCardView staxCardView;
		private LinearLayout bountyListView;

		public CardedBountyListViewHolder(@NonNull View itemView) {
			super(itemView);
			staxCardView = itemView.findViewById(R.id.bountyChannelCard);
			bountyListView = itemView.findViewById(R.id.bountyList);
		}
	}

	@Override
	public long getItemId(int position) {
		return position;
	}

	@Override
	public int getItemViewType(int position) {
		return position;
	}

}

File: app/src/main/java/com/hover/stax/bounties/BountyChannelsAdapter.kt
package com.hover.stax.bounties

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.hover.stax.bounties.BountyChannelsAdapter.CardedBountyListViewHolder
import com.hover.stax.channels.Channel
import com.hover.stax.databinding.BountyCardChannelBinding

internal class BountyChannelsAdapter(private var channelList: List<Channel>?, private val allBountiesList: List<Bounty>, private val selectListener: BountyListItem.SelectListener) : RecyclerView.Adapter<CardedBountyListViewHolder>() {

    init {
        channelList = channelList?.asSequence()?.filter { filterBounties(it.id).any { b -> b.action.bounty_is_open || b.transactionCount != 0 } }?.toList()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CardedBountyListViewHolder {
        val binding = BountyCardChannelBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return CardedBountyListViewHolder(binding)
    }

    override fun onBindViewHolder(holder: CardedBountyListViewHolder, position: Int) {
        val c = channelList!![position]
        holder.binding.bountyChannelCard.setTitle(c.ussdName)
        val channelBounties = filterBounties(c.id)
        for (b in channelBounties) {
            if (b.action.bounty_is_open || b.transactionCount != 0) {
                val bountyLi = BountyListItem(holder.binding.bountyChannelCard.context, null)
                bountyLi.setBounty(b, selectListener)
                holder.binding.bountyList.addView(bountyLi)
            }
        }
    }

    private fun filterBounties(channelId: Int): List<Bounty> {
        val matches: MutableList<Bounty> = ArrayList()
        for (b in allBountiesList) {
            if (b.action.channel_id == channelId) {
                matches.add(b)
            }
        }
        return matches
    }

    override fun getItemCount(): Int {
        return channelList?.size ?: 0
    }

    internal class CardedBountyListViewHolder(var binding: BountyCardChannelBinding) : RecyclerView.ViewHolder(binding.root)

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getItemViewType(position: Int): Int {
        return position
    }
}